proxylib fixes

jrajahalme opened the pull request:
Address the remaining review comments on #5174

Allowing an L7 proto to be specified without rules turned out to be more complicated than maybe expected.

Also found out that L7 rule merging was done for ingress only, refactored the code to remove duplication between ingress/egress.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5635)
<!-- Reviewable:end -->

tgraf commented on proxylib/npds/client.go:
```diff
@@ -20,6 +20,7 @@ import (
 	"io"
 	"time"
 
+	"github.com/cilium/cilium/pkg/backoff"
```
Add a fixes tag to the commit message for the issue that I had opened.

jrajahalme replied:
Done.

tgraf commented on proxylib/proxylib/connection.go:
```diff
@@ -30,9 +30,7 @@ import (
-type Direction struct {
-	InjectBuf *[]byte
-}
+type InjectBuf *[]byte
```
Please add a description to explain the details on why this is a pointer to a byte slice

jrajahalme replied:
Added this:
```
// InjectBuf is a pointer to a slice header for an inject buffer allocated by
// the proxylib caller. As data is placed into the buffer, the length
// of the buffer in the slice header is increased correspondingly. To make
// the the injected data visible to the caller we need to pass the slice header
// by reference rather than by value, hence the pointer in the type.
// As the caller is typically in a differnent memory management domain (not
// subject to Go runtime garbage collection), the underlying buffer may never
// be expanded or otherwise reallocated.
```

jrajahalme replied:
I used the same opportunity to comment the Connection struct and its fields.

jrajahalme committed:
proxylib/npds: Use pkg/backoff

Use pkg/backoff instead of hand-rolling exponential backoff for NPDS
client connection attempts from proxylib.

Fixes: #5618
Suggested-by: Thomas Graf <thomas@cilium.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxylib/npds: Recognize io.ErrUnexpectedEOF

Unexpected EOFs are to be expected when connections are being torn
down. Do not log them as errors.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxylib: Flatten 'Direction' to a 'InjectBuf' type.

Since the 'Direction' struct only has the inject buffer pointer,
define a type for the inject buffer pointer and use it instead.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
api: Allow missing L7 rules

Missing (nil) L7 rules should behave the same as an empty set of rules
when 'l7proto' has been specified.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy/api: Add PortRuleL7 unit tests.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
pkg/policy: Fix L7 policy merging.

The generic L7 policy merging logic was missing from the Egress case
and the Ingress case was also not completely correct.

Factor out the common code to a new mergeL4Port() helper and test for
the presence of the 'L7Proto' field rather than an empty set of L7
rules, so that conflict is detected also when the list of L7 rules is
empty, but the L7Proto is specified.

Require the L7Proto field also in unit testing.

Detect merge conflicts due to a specified 'L7Proto' even when there are no rules.

Add a L7 policy merge test case also for egress.

Print out L7 and Kafka rules on policy trace.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
k8s: Loosen the language on PortRuleL7 description.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
.gitignore: Ignore envoy/proxylib

The header files for proxylib must be copied to the envoy directory
for Bazel to be be able to refer to them. Same applies for the library
itself for the purpose of Envoy integration testing. 'envoy/proxylib'
is created for this purposes when running make. Make git ignore it.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
proxylib: Factor out logic scanning the input slices.

Factor out the logic scanning the input slices to advanceInput() and
add unit testing for it.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
